<a href="https://colab.research.google.com/github/patiltejas20/ML-basic-project/blob/main/ML_mini_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import numpy as np

# Generate some sample data
X = np.random.rand(100, 5)  # 100 samples, 5 features
y = np.random.randint(0, 2, 100)  # 100 labels (0 or 1)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a KNN classifier with k=3
knn = KNeighborsClassifier(n_neighbors=3)

# Train the model
knn.fit(X_train, y_train)

# Make predictions on the test set
y_pred = knn.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

Accuracy: 0.4


In [2]:
# Train the model
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

# Task
Create a mini-project that generates random data, preprocesses it, and trains a machine learning model.

## Generate sample data

### Subtask:
Create a synthetic dataset with multiple features and a target variable.


**Reasoning**:
Generate a synthetic dataset with 1000 samples, 20 features (15 informative, 5 redundant), and 2 classes using `make_classification`. Store the features in a variable named `X` and the target variable in a variable named `y`.



In [4]:
from sklearn.datasets import make_classification

X, y = make_classification(n_samples=1000, n_features=20, n_informative=15, n_redundant=5, n_classes=2, random_state=42)

## Preprocess data

### Subtask:
Handle potential issues in the data, such as scaling numerical features.


**Reasoning**:
Import the StandardScaler, instantiate it, fit it to the data, and transform the data.



In [6]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X)
X_scaled = scaler.transform(X)

## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the scaled feature data and the target variable into training and testing sets using train_test_split with a test size of 0.2 and a random state of 42.



In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

## Train ml model

### Subtask:
Choose and train an appropriate machine learning model on the training data.


**Reasoning**:
Import LogisticRegression, instantiate it, and train the model.



In [10]:
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=42)
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=42)

## Evaluate model

### Subtask:
Evaluate the performance of the trained model on the testing data.


**Reasoning**:
Make predictions using the trained model and evaluate the accuracy.



In [12]:
y_pred = logistic_regression_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

Accuracy: 0.825


## Summary:

### Data Analysis Key Findings

*   A synthetic dataset was generated with 1000 samples, 20 features (15 informative, 5 redundant), and 2 classes.
*   The numerical features of the synthetic data were successfully scaled using `StandardScaler`.
*   The scaled data was split into training and testing sets, with 80% for training and 20% for testing.
*   A Logistic Regression model was trained on the training data.
*   The trained Logistic Regression model achieved an accuracy of 0.825 on the test data.

### Insights or Next Steps

*   The model's performance of 82.5% accuracy on the test set suggests it is capable of making reasonable predictions on unseen data.
*   Further steps could involve exploring other classification models, hyperparameter tuning, or evaluating other metrics like precision, recall, or F1-score to gain a more comprehensive understanding of the model's performance.
